contracts/src/Gateway.sol

```diff
@@ -410,6 +415,35 @@ contract Gateway is IGateway, IInitializable {
         _transferNativeFromAgent(agent, payable(params.recipient), params.amount);
     }
 
+    struct RegisterTokenParams {
+        /// @dev The ID of the agent to register the token for
+        bytes32 agentID;
+        /// @dev The name of the new token
+        string name;
+        /// @dev The short symbol of the new token
+        string symbol;
+        /// @dev The number of decimal places allowed
+        uint8 decimals;
+    }
+
+    // @dev Register a new fungible Polkadot token for an agent
+    function registerToken(bytes calldata data) external onlySelf {
+        CoreStorage.Layout storage $ = CoreStorage.layout();
+
+        RegisterTokenParams memory params = abi.decode(data, (RegisterTokenParams));
+
+        address agent = $.agents[params.agentID];
+        if (agent == address(0)) {
+            revert AgentDoesNotExist();
+        }
+
+        bytes memory call = abi.encodeCall(AgentExecutor.registerToken, (params.name, params.symbol, params.decimals));
+        bytes memory response = _invokeOnAgent(agent, call);
+
+        address token = address(bytes20(response));
+        $.ownerAgentIDs[token] = params.agentID;
+    }
```

Comment:
The gateway already has an `Command.AgentExecute` command for sending arbitrary messages to Agents. Lets use that rather than adding special-purpose commands for higher-level application stuff. 

I understand we do need to keep a token registry in the gateway. So you could make the agent call back into the gateway to update the registry.

Reply:
Oops, my mistake, your approach here is correct. My memory was foggy as I just got back from leave.

Reply:
Actually, yes, we should be reusing using 'Command.AgentExecute' instead of creating a new toplevel 'Command.RegisterToken'.

Reply:
The call back into the gateway opens a door to attackers: they can change the agentID for any token. This won't let them control the token (as they're not the owner recorded in the token contract), but it can be used to grief legitimate owners by changing the token's agentID so that calls to the token fail.

So we'll need some kind of authorisation when updating the token registry. It might be enough to require that the token hasn't been registered before.